db/db_versioned.go

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) 2024 IoTeX Foundation
+// This is an alpha (internal) release and is not suitable for production. This source code is provided 'as is' and no
+// warranties are given as to title or non-infringement, merchantability or fitness for purpose and, to the extent
+// permitted by law, all liability for your use of the code is disclaimed. This source code is governed by Apache
+// License 2.0 that can be found in the LICENSE file.
+
+package db
+
+type (
+	// KvVersioned is a versioned key-value store, where each key has multiple
+	// versions of value (corresponding to different heights in a blockchain)
+	//
+	// Versioning is achieved by using (key + 8-byte version) as the actual
+	// storage key in the underlying DB. For each bucket, a metadata is stored
+	// at the special key = []byte{0}. The metadata specifies the bucket's name
+	// and the key length.
+	//
+	// For each versioned key, the special location = key + []byte{0} stores the
+	// key's metadata, which includes the following info:
+	// 1. the version when the key is first created
+	// 2. the version when the key is lastly written
+	// 3. the version when the key is deleted
+	// 4. hash of the key's last written value (to detect/avoid same write)
+	// If the location does not store a value, the key has never been written.
+	//
+	// How to use a versioned DB:
+	//
+	// db := NewBoltDBVersioned(cfg) // creates a versioned DB
+	// kv := db.SetVersion(5)
+	// value, err := kv.Get("ns", key) // read 'key' at version 5
+	// kv = db.SetVersion(8)
+	// err := kv.Put("ns", key, value) // write 'key' at version 8
+
+	KvVersioned interface {
+		// Version returns the key's most recent version
+		Version(string, []byte) (uint64, error)
+
+		// SetVersion sets the version, and returns a KVStore to call Put()/Get()
+		SetVersion(uint64) KVStore
+	}
+
+	// BoltDBVersioned is KvVersioned implementation based on bolt DB
+	BoltDBVersioned struct {
+		*BoltDB
+		version uint64 // version for Get/Put()
+	}
+)
+
+// Option sets an option
+type Option func(b *BoltDBVersioned)
+
+// NewBoltDBVersioned instantiates an BoltDB which implements KvVersioned
+func NewBoltDBVersioned(cfg Config, opts ...Option) *BoltDBVersioned {
+	b := &BoltDBVersioned{
+		BoltDB: NewBoltDB(cfg),
+	}
+	for _, opt := range opts {
+		opt(b)
+	}
+	return b
+}
+
+// Put writes a <key, value> record
+func (b *BoltDBVersioned) Put(ns string, key, value []byte) error {
+	if !b.IsReady() {
+		return ErrDBNotStarted
+	}
+	// TODO: implement Put
+	return nil
+}
+
+// Get retrieves the most recent version
+func (b *BoltDBVersioned) Get(ns string, key []byte) ([]byte, error) {
+	if !b.IsReady() {
+		return nil, ErrDBNotStarted
+	}
+	// TODO: implement Get
+	return nil, nil
+}
+
+// Delete deletes a record,if key is nil,this will delete the whole bucket
+func (b *BoltDBVersioned) Delete(ns string, key []byte) error {
+	if !b.IsReady() {
+		return ErrDBNotStarted
+	}
+	// TODO: implement Delete
+	return nil
+}
```

Comment:
Do we still need `Put`, `Get`, `Delete` the three methods?

Reply:
yes, that is the core function of this `KvVersioned` concept, and will be implemented in later PRs